HTTP, SMTP, and DNS parser enhancement

codesniffer opened the pull request:
1. HTTP parser can parse all the request and response headers.

2. SMTP parser can parse all the headers

3. DNS parser can parse and store authoritative and additional IPs and hostnames

awick commented on capture/parsers/http.c:
```diff
@@ -332,6 +340,28 @@ LOCAL int moloch_hp_cb_on_header_value (http_parser *parser, const char *at, siz
+/*------------VISA Enhancement Begin---------------*/
+    if(http->which == 0) { // Headers in HTTP request
+        if (config.parseHTTPHeaderRequestAll) {
+            char *req_header_field = g_strdup(http->header[http->which]);
```
There is no reason to do a strdup, and then a copy, and then a free.  
            moloch_field_string_add(headerReqField, session, http->header[http->which], -1, TRUE);
            moloch_field_string_add(headerReqValue, session, at, length, TRUE);

same for the below.  BTW, if the value is across calls this will not work since you should be concating the values like the inValue stuff does.

awick commented on capture/config.c:
```diff
@@ -535,9 +535,12 @@ void moloch_config_load()
     config.logFileCreation       = moloch_config_boolean(keyfile, "logFileCreation", config.debug);
     config.logHTTPConnections    = moloch_config_boolean(keyfile, "logHTTPConnections", TRUE);
     config.parseSMTP             = moloch_config_boolean(keyfile, "parseSMTP", TRUE);
+    config.parseSMTPHeaderAll    = moloch_config_boolean(keyfile, "parseSMTPHeaderAll", TRUE);
```
Would like to default all 3 to false since this will add a lot of data to ES

awick commented on capture/parsers/dns.c:
```diff
@@ -27,6 +27,119 @@ LOCAL  int                   queryClassField;
+
+/*------------VISA Enhancement Begin---------------*/
+
+/* Reference https://github.com/spc476/SPCDNS/blob/master/src/dns.h */
```
We don't allow any LGPL v3 or GPL v3, you will have to find another source

awick commented on capture/parsers/dns.c:
```diff
@@ -139,14 +252,40 @@ LOCAL void dns_parser(MolochSession_t *session, int kind, const unsigned char *d
+    resultRecordCount [1] = nscount;
+    resultRecordCount [2] = arcount;
+
+    int _ipField = 0;
```
No _ variable names please, C standard

awick commented on capture/parsers/dns.c:
```diff
@@ -207,8 +350,26 @@ LOCAL void dns_parser(MolochSession_t *session, int kind, const unsigned char *d
+      
+      for (i = 0; BSB_NOT_ERROR(bsb) && i < recordNum; i++) {
+
+          if (recordType == RESULT_RECORD_ANSWER) {
```
switch please

awick commented on capture/parsers/dns.detail.jade:
```diff
@@ -1,8 +1,16 @@
+    +ipArrayListDNS(session.dns, "update", "IPs Update","dns.update")  
+    +arrayList(session.dns, "query.host","Hosts Query", "host.dns.query")
+    +arrayList(session.dns, "answer.host","Hosts Answer","host.dns.answer")
+    +arrayList(session.dns, "authoritative.host","Hosts Authoritative","host.dns.authritative")
```
typo

awick commented on capture/parsers/smtp.c:
```diff
@@ -544,6 +549,20 @@ LOCAL int smtp_parser(MolochSession_t *session, void *uw, const unsigned char *d
+            if (config.parseSMTPHeaderAll) {
+                int colon_position = colon - line->str + 1;
+                char *smtp_header_field = g_strndup(line->str, colon - line->str);
+                char *smtp_header_value = g_strndup(line->str + colon_position, line->len - colon_position); 
```
don't need the strdups

awick commented on viewer/views/mixins.pug:
```diff
@@ -142,3 +142,45 @@ mixin ipArrayList(container, field, title, expr, max)
         |</span>
         a.cursor-pointer.show-more-items &hellip;
+
+mixin ipArrayListDNS(container, field, title, expr, max)
```
any reason not to have this in the dns.jade file, its not reusable

awick commented:
1) Please add to the tests/config.test.ini the end of the default section and add to pr
```
parseSMTPHeaderAll=true
parseHTTPHeaderRequestAll=true
parseHTTPHeaderResponseAll=true
```
2) Run tests.pl in there, issues I notice

* pcap/wireshark-bdat  the white space before the headerValues should be removed
```
# *  39|          headerValue => [                                                                                                                                                  *    |                                 |
# *  40|            ' from root by testhost.test.ex with local-bsmtp (Exim x.yz) (envelope-from <sender@dom>) id 1bwpfB-0000Be-Tu for p@test.ex; Wed, 19 Oct 2016 13:08:05 +0100',  *    |                                 |
# *  41|            ' foo',                                                                                                                                                         *    |                                 |
# *  42|            ' <E1bwpfB-0000Be-Tu@testhost.test.ex>',                                                                                                                        *    |                                 |
# *  43|            ' sender@dom',                                                                                                                                                  *    |                                 |
# *  44|            ' Wed, 19 Oct 2016 13:08:05 +0100'                                                                                                                              *    |                                 |
# *  45|          ],                                                                                                                                                                *    |                                 |
```

* I just want to confirm for http headers you are excluding ones already indexed, such has host/user-agent

* Is there really a need for two counts?  responseHeaderValueCnt and responseHeaderFieldCnt should always be the same right?

* For DNS it looks like the field name for some results has moved from ip to answer.ip.  Not sure we should be renaming fields.  I need to look at DNS more.

3) For tests you are happy with run `./tests.pl --make pcap/file.pcap` and add to the PR

codesniffer commented:
**Q.** I just want to confirm for http headers you are excluding ones already indexed, such has host/user-agent
**A**. Yes.

 
**Q.** Is there really a need for two counts? responseHeaderValueCnt and responseHeaderFieldCnt should always be the same right?
**A.** Yes, the values of these two fields should always be the same. The requestHeaderFieldCnt and responseHeaderFieldCnt  fields can be removed. Please advise

**Q.** For DNS it looks like the field name for some results has moved from ip to answer.ip. Not sure we should be renaming fields. I need to look at DNS more.
**A.**  The dns.ip and dns.host contain all the ips and host name. I added separate fields to clearly show what IP and host belong to what resource record (query, answer, additional, and authoritative). I introduced an option in config.ini (parseDNSRecordsAll) to control parsing of DNS metadata. If parseDNSRecordsAll is set to TRUE then fields for query, answer, additional, and authoritative will be populated and shown in the UI. I am showing the IP and host break down in a separate section (DNS Record Details) in the UI. Please review and advise

**Q.** For tests you are happy with run ./tests.pl --make pcap/file.pcap and add to the PR
**A.** I made changes in capture test castes according to your suggestion. The capture test-cases are passing. However, some of the viewer test cases are failing. I am working to fix those.

awick commented:
Great, making progress.  
* Can you make sure there are no trailing spaces in new code "/ $", git will highlight
* I'm not sure I want the new hosts/ips added to old dns fields.  Is that just to make queries easier?

codesniffer commented:
Q. Can you make sure there are no trailing spaces in new code "/ $", git will highlight
A. Sure, I will work on it.

Q. I'm not sure I want the new hosts/ips added to old dns fields. Is that just to make queries easier?
A. Yes, one reason is to make the query easier. A user can search for any host and ip using **dns.ip** and **dns.host** APIs.  
I think this is a design problem. I need your suggestion in this regard. When parseDNSRecordAll is set TRUE, then I populate the IPs  _(dns.query.ip, dns.answer.ip, dns.authoritative.ip, dns.additional.ip, dns.update.ip)_ and hosts _( dns.query.host, dns.answer.host, dns.authoritative.host, dns.additional.host, dns.update.host)_  from DNS records to show a break down in the UI. At the same time, I populate the _dns.host = dns.query.host U dns.answer.host U dns.authoritative.host, dns.additional.host U dns.update.host_ and _dns.ip = dns.query.ipU dns.answer.ip U dns.authoritative.ip, dns.additional.ip U dns.update.ip._ The dns.host and dns.ip are shown in the DNS section, which means the DNS section shows a unified information. Next, I show the IP and Host break down in the DNS Record Detail section.

awick commented:
I think probably shouldn't populate it twice, instead should maybe make two new fake fields that query all those.  like host.dns.all and ip.dns.all maybe?  And leave host.dns and ip.dns the way it is now.  Or maybe even turn those off writing ip.dns and host.dns fields.  duplicate data in es always makes me sad.  speaking of duplicate data, since you don't duplicate http headers, it looks like you are duplicate smtp headers.  on purpose?

codesniffer commented:
Yes, I am parsing SMTP header twice. I did not notice it. Thanks for pointing it out. I will provide a fix in the next commit. The next commits will include the following changes: a) SMTP header parsing fix; b) DNS header fix c) viewer taste cases fix. Please suggest.

codesniffer commented:
I have completed making the changes. The changes are as follows:

**HTTP Parser**
1. If _parseHTTPRequestHeaderAll_ is set to _TRUE_ in the config.ini then all the HTTP-headers are parsed and stored in ES. 
2. If parseHTTPResponseHeaderAll is set to _TRUE_ in the config.ini, then all the HTTP-headers are parsed and stored in ES. 
3. It is ensured that no HTTP header is parsed twice
4. The values of the HTTP headers are shown in a separate section (HTTP Header Detail) in the session detail GUI
5. Test cases were updated according to the changes

**SMTP Parser**

1. If _parseSMTPHeaderAll_ is set to _TRUE_ in the config.ini, then all the email-headers are parsed and stored in ES. It is ensured that no email header is parsed twice
3. The values of the Email headers are shown in a separate section (Email Header Detail) in the session detail GUI
4. SMTP test cases were updated according to the changes

**DNS Parser**

1. If _parseDNSRecordAll_ is set to _TRUE_ in the config.ini then results from Authoritative and Additional records are parsed. Name Server and Mail Exchange server information (IPs and Hostnames) are in separate labels in the UI.
2. DNS test cases were updated according to the changes
